### PR TITLE
azure - Add requests limit logs & custom retry code

### PR DESCRIPTION
**Pull request by logachev**
- Log x-ms-ratelimit* headers
- Custom retry code to avoid urllib3 waiting for the Retry-After value

### CODE REVIEW

File: tools/c7n_azure/c7n_azure/utils.py [hunk 1]
```diff
@@ -99,6 +100,41 @@ def azure_name_value_pair(name, value):
...
+    while retries < max_retries:
+        response = self.old_send(request, headers, content, **kwargs)
+
+        send_logger.debug(response.status_code)
+        for k, v in response.headers.items():
+            if k.startswith('x-ms-ratelimit'):
+                send_logger.debug(k + ':' + v)
+
+        # Retry codes from urllib3/util/retry.py
+        if response.status_code in [413, 429, 503]:
```
Comment (stefangordon) [inline tools/c7n_azure/c7n_azure/utils.py]: Do you think this merits a warning level log? Just thinking about how we can give ourselves visibility into when this is happening to help plan future enhancements here.

File: tools/c7n_azure/c7n_azure/utils.py [hunk 2]
```diff
@@ -99,6 +100,41 @@ def azure_name_value_pair(name, value):
...
+            retry_after = None
+            for k in response.headers.keys():
+                if StringUtils.equal('retry-after', k):
+                    retry_after = int(response.headers[k])
+
+            if retry_after is not None and retry_after < 30:
```
Comment (aluong) [inline tools/c7n_azure/c7n_azure/utils.py]: (minor) lift the 30 seconds up to a var, so we can parameterize this in the future easier.

**Comment by aluong** [comment]
Can we add some more details in the PR description on the existing behavior and the new behavior you are adding?

**Comment by aluong** [comment]
Actually, can we look into subclassing the retry urlib3 retry class and adding custom retry logic there oppose to patching?

In addition, it looks like you can add that custom logging you want by modifying the pipeline used by AutoRest.
https://github.com/Azure/msrest-for-python/wiki/msrest-0.6.0---Pipeline

**Pull request merged**